src/core/device.c

```diff
@@ -349,6 +361,29 @@ static int device_setup_unit(Manager *m, struct udev_device *dev, const char *pa
                         (void) device_add_udev_wants(u, dev);
         }
 
+        DEVICE(u)->bind_mounts = device_is_bound_by_mounts(dev);
+
+        /* So the user wants the mount units to be bound to the device but a
+         * mount unit might has been seen by systemd before the device appears
+         * on its radar. In this case the device unit is partially initialized
+         * and includes the deps on the mount unit but at that time the "bind
+         * mounts" flag wasn't not present. Fix this up now. */
+        if (DEVICE(u)->bind_mounts) {
+                Unit *other;
+                Iterator i;
+
+                SET_FOREACH(other, u->dependencies[UNIT_REQUIRED_BY], i) {
+                        if (other->type != UNIT_MOUNT)
+                                continue;
+
+                        if (set_get(u->dependencies[UNIT_BOUND_BY], other))
+                                continue;
+
+                        r = unit_add_dependency(other, UNIT_BINDS_TO, u, true);
+                        if (r < 0)
+                                goto fail;
+                }
```

Comment:
i think this new loop should probably be its own function "device_maybe_update_deps()" or so?